test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -257,3 +297,33 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
 
   // Communicators might be aborted here, further operations would fail.
 }
+
+TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNoHeartbeat) {
+  if (skipTest()) {
+    return;
+  }
+
+  ASSERT_TRUE(setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1) == 0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT, "3", 1) == 0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  auto options = c10d::ProcessGroupNCCL::Options::create();
+  options->timeout = std::chrono::milliseconds(30000);
+  ProcessGroupNCCLNoHeartbeat pg(store_, 0, 1, options);
+
+  auto work = pg.allreduce(tensors_);
+  work->wait();
+  EXPECT_TRUE(work->isSuccess());
+
+  work = pg.allreduce(tensors_);
+  {
+    // Now run all reduce with errors.
+    std::lock_guard<std::mutex> lock(pg.getWatchdogMutex());
+    LOG(INFO) << "Lock watchdog thread.";
+    // Wait for a while before monitor thread throws exceptions.
+    std::this_thread::sleep_for(std::chrono::seconds(10));
```

Comment:
2 suggestions
 - don't hardcode 10 sec here, its better if you compute 2 * heartbeat_interval or something like that
 - maybe for this test case you can even configure the heartbeat interval to be fast, then reduce the test time from 10sec to 1sec.  (just from annoyance, it gets tiring for a test to be slow when you need to repro an issue later)

Reply:
Well. because `ProcessGroupNCCL::kWatchdogThreadSleepMillis = 1000`, so heartbeat check needs to be at least every 2 secs. That's why I use 10 seconds.. Let me see if I can make it shorter.